tlatools/org.lamport.tlatools/src/tlc2/tool/impl/Tool.java

```diff
@@ -3486,9 +3486,112 @@ else if (val == null) {
         return this.enabled(acts, s0, s1, cm);
   }
 
+    @Override
+    public InvariantViolation findFalsifyingContext(Action invariant, TLCState s0, TLCState s1) {
+      return findFalsifyingContext(
+              invariant.pred,
+              invariant.con,
+              s0,
+              s1,
+              invariant.cm,
+              invariant.con);
+    }
+
+    /**
+     * Helper for {@link ITool#findFalsifyingContext(Action, TLCState, TLCState)}.
+     */
+    private InvariantViolation findFalsifyingContext(SemanticNode expr, Context c, TLCState s0, TLCState s1, CostModel cm, Context relevantAssignments) {
```

Comment:
I didn't review the logic, assuming it has been largely copied from `isValid`?!

Reply:
Well, mostly copied from `evalApplImpl` (which is what you get if you inline a few methods). But essentially yes. :)